zcash_primitives/src/note_encryption.rs

```diff
@@ -0,0 +1,1079 @@
+/// let outCiphertext = enc.encrypt_outgoing_plaintext(&cv.cm(&JUBJUB).into(), &cmu);
+/// ```
+pub struct SaplingNoteEncryption {
+    epk: edwards::Point<Bls12, PrimeOrder>,
+    esk: Fs,
+    note: Note<Bls12>,
+    to: PaymentAddress<Bls12>,
+    memo: Memo,
+    ovk: OutgoingViewingKey,
+}
+
+impl SaplingNoteEncryption {
+    /// Creates a new encryption context for the given note.
+    pub fn new(
+        ovk: OutgoingViewingKey,
+        note: Note<Bls12>,
+        to: PaymentAddress<Bls12>,
+        memo: Memo,
+    ) -> SaplingNoteEncryption {
+        let esk = generate_esk();
+        let epk = note.g_d.mul(esk, &JUBJUB);
+
+        SaplingNoteEncryption {
+            epk,
+            esk,
+            note,
+            to,
+            memo,
+            ovk,
+        }
+    }
+
+    /// Exposes the ephemeral secret key being used to encrypt this note.
+    pub fn esk(&self) -> &Fs {
+        &self.esk
+    }
+
+    /// Exposes the ephemeral public key being used to encrypt this note.
+    pub fn epk(&self) -> &edwards::Point<Bls12, PrimeOrder> {
+        &self.epk
+    }
+
+    /// Generates `encCiphertext` for this note.
+    pub fn encrypt_note_plaintext(&self) -> [u8; ENC_CIPHERTEXT_SIZE] {
+        let shared_secret = sapling_ka_agree(&self.esk, &self.to.pk_d);
+        let key = kdf_sapling(&shared_secret, &self.epk);
+
+        let mut input = Vec::with_capacity(NOTE_PLAINTEXT_SIZE);
+        input.push(1);
+        input.extend_from_slice(&self.to.diversifier.0);
+        (&mut input)
+            .write_u64::<LittleEndian>(self.note.value)
+            .unwrap();
+        self.note.r.into_repr().write_le(&mut input).unwrap();
+        input.extend_from_slice(&self.memo.0);
+        assert_eq!(input.len(), NOTE_PLAINTEXT_SIZE);
+
+        let mut output = [0u8; ENC_CIPHERTEXT_SIZE];
+        assert_eq!(
+            ChachaPolyIetf::aead_cipher()
+                .seal_to(&mut output, &input, &[], &key.as_bytes(), &[0u8; 12])
+                .unwrap(),
+            ENC_CIPHERTEXT_SIZE
+        );
+
+        output
+    }
+
+    /// Generates `outCiphertext` for this note.
+    pub fn encrypt_outgoing_plaintext(
+        &self,
+        cv: &edwards::Point<Bls12, Unknown>,
+        cmu: &Fr,
+    ) -> [u8; OUT_CIPHERTEXT_SIZE] {
+        let key = prf_ock(&self.ovk, &cv, &cmu, &self.epk);
+
+        let mut buf = [0u8; OUT_CIPHERTEXT_SIZE];
+        self.note.pk_d.write(&mut buf[0..32]).unwrap();
+        self.esk
+            .into_repr()
+            .write_le(&mut buf[32..OUT_PLAINTEXT_SIZE])
+            .unwrap();
+
+        assert_eq!(
+            ChachaPolyIetf::aead_cipher()
+                .seal(
+                    &mut buf,
+                    OUT_PLAINTEXT_SIZE,
+                    &[],
+                    key.as_bytes(),
+                    &[0u8; 12]
+                )
+                .unwrap(),
+            OUT_CIPHERTEXT_SIZE
+        );
+        buf
```

Comment:
Opened https://github.com/KizzyCode/crypto_api_chachapoly/issues/4

Reply:
@daira This is a good point. Currently, the entire ChaCha20 implementation works by creating a key-stream block and XORing this block directly into the target buffer because this is more efficient than creating a key-stream block, XORing this block in-place with a source block and copying the result into the target buffer.
Currently I'm implementing the changes suggested by @str4d (the `seal_to`/`open_to` functions will have their own checks) – I'll keep your point in mind for future development 😊